File: problem_040.c
// -*- compile-command: "gcc -o problem_040 problem_040.c ceuler.c -Wall -lm" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include "ceuler.h"

int main()
{
    printf("Project Euler - Problem 40:\n"
           "If dn represents the nth digit of the fractional part, find the value of the following expression.\n"
           "d1 × d10 × d100 × d1000 × d10000 × d100000 × d1000000 \n\n");

    unsigned long number = 1;
    unsigned long count  = 0;
    int fraction[1000100]={};

    // Fill up array with sequential integers
    while (count < 1000000)
    {
        unsigned long tmpNum   = number;
        unsigned long tmpCount = numberOfDigits(number);
        // Add digit to array in reverse order
        while (tmpCount > 0)
        {
            fraction[count+tmpCount] = tmpNum % 10;
            tmpNum /= 10;
            tmpCount--;
        }
        count += numberOfDigits(number);
        number++;
    }

    // Perform the multiplication
    number = fraction[1] *    fraction[10] *    fraction[100] *
             fraction[1000] * fraction[10000] * fraction[100000] *
             fraction[1000000];

    printf("Total: %lu\n", number);

    return 0;
}

File: problem_040.swift
#!/usr/bin/xcrun swift
// -*- compile-command: "swiftc problem_040.swift -o problem_040" -*-
// Copyright (c) 2016 Michael Caldwell
import Cocoa

print("Project Euler - Problem 40:")
print("If dn represents the nth digit of the fractional part, find the value of the following expression.\nd1 × d10 × d100 × d1000 × d10000 × d100000 × d1000000 \n")

var array = [0]
for n in 1...1000000 {
    let l = Int(floor(log10(Float(n))))
    for i in (0...l).reverse() {
        var t = n / Int(pow(Float(10), Float(i)))
        t = t % 10
        array.append(t)
    }
    if array.count > 1000000 {
        break
    }
}

var answer = 1
for i in 1...6 {
    answer = answer * array[Int(pow(10.0, Float(i)))]
}

print(answer)
